---
title: "Lab 1 - Intro - CB2040"
author: ""
date: "15 Sep 19"
output:
  BiocStyle::html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

# Lab 1

The aim of this first lab is to get you familiar with working with R and to create R Markdown files, i.e. there will not be much exciting happening here in terms of data analysis, more to get you into the workflow.
For all the labs, we will want you to create a R markdown report to hand in at the end of the lab course, i.e. it's good to be familiar with this type of document ;) 

Note, you will **not** have to hand in lab1.

Note, it's totally fine to do this lab wherever/whenever you want, at the scheduled lab occasion, at home, on the bus (?). Just hand it in before the deadline. But keep in mind that if you want any help at any stage, please come to the lab occasion.

For people with previous knowledge of R, congrats! You will be able to finish this first lab in no time :)

# Install R (skip if already done)

Follow the [link](https://cran.r-project.org/) and then choose the operational system you use.

- Choose the option ‘install R for the first time’ and download the setup file.
- Open the downloaded file and Install R.
- Run the installation file and you will be able to see the standard version of the console.

## Install RStudio

RStudio is an integrated development environment for R, in simple terms, a code editor.

- Follow the [link](https://www.rstudio.com/products/rstudio/#Desktop) and click the Download RStudio Desktop button.
- Open the downloaded .exe file and Install RStudio.
- After RStudio is installed, find and launch it.

# Intro To R Markdown

Open a new file by File -> New File -> R Markdown..

When you open this file, you will get open an R Markdown template. Proceed with this lab by typing everything within this document. Lets begin by going through some basic R functions. 

Use [this cheat sheet](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf) and follow the step guide (1-6) to understand the structure of the R Markdown document, you can also use this sheet as a go-to when you look for R Markdown commands, e.g. how to create headers, bold text, Italic text, bullet point lists etc. 

Try to start with a clean sheet, write the code as instructed in this lab, and finally "Knit" the document to a .html file. This is the same output format we will use for the hand in of lab 2-4 (.html or .pdf).

# Basic R operations

## Look for help

For all functions in R, the easiest way to find how to use them is simply by typing `?[the name of the function]`, e.g. `?sum()`. Which will give you the explanation for the function!

## Basic arithmetic operations

```{r, eval=FALSE}
7 + 4 # => 11
7 - 4 # => 3
7 / 2 # => 3.5
7 * 2 # => 14
```

## Basic arithmetic functions


```{r, eval=FALSE}
log2(4) # => 2
abs(-4) # => 4
sqrt(4) # => 2
```

## Objects

To create new variables, you will need to use the assignment operator (<-) or (=)

```{r}

object <- 5
object = 5

print(object)

```

### Vectors

Create a vector

```{r}

# Creating a vector
vector <- c(1, 19, 7 + 1, 123)
vector

#With characters
example <- c("GeneA", "GeneB")
example

```

Calculations with vectors

max(x), min(x), range(x), length(x), sum(x), mean(x)

```{r}

max(vector)

range(vector)

sum(vector)


```


### Data frames

Data frames looks like a matrix but can have columns with different types. 

Create a data frame

```{r}

# Create the data frame.
count.data <- data.frame(
   genes = c("A","B","C","D"),
   counts = c(5, 10, 1, 30)
)
# Print the data frame.			
print(count.data) 


```

Our data frame now has "column"-names "genes" and "counts".

```{r}
colnames(count.data)

```

Lets make a data frame of several samples in the columns, and instead of having a specific column for the gene names, lets put those as "rownames" instead.

```{r}

count.data <- data.frame(
   sample1 = c(2, 3, 4, 50),
   sample2 = c(5, 10, 1, 30)
)
rownames(count.data) <- c("A", "B", "C", "D")
count.data

```

You can access the data frame in different ways. 
To view sample1:

```{r}

#data.frame$"column-name"

count.data$sample1

```

or, the same thing would be to view all rows in the first column:

```{r}

#Access [rows, columns]

count.data[, 1]

```

Calculations with data frames

Calculations with numeric data frames: rowSums(), colSums(), rowMeans(), colMeans()

```{r}

rowSums(count.data)

colMeans(count.data)


```

Subset a data frame and perform calculations on that

```{r}

count.data[, 2] = count.data[, 2] + 2

count.data

```

Add to a data frame

You can add a row with `rbind()`

```{r}

new.row = c(2, 4)
count.data = rbind(count.data, E=new.row)

count.data

```

Similar, you can use `cbind()` to add a column, or, add via the `$` operator

```{r}

count.data$sample3 = c(1,2,3,4,5)

count.data

```

Now, create a count table of 3 samples and 5 genes. Make all count values = 1

```{r}


```

Make sample 2 to have double amount of counts for each gene

```{r}


```

Make sample 3 to have 5x counts for each gene

```{r}


```

Cut the count values for gene 2 in half, and add +3 to gene 4 for each sample

```{r}


```


Calculate the mean value of each gene and store this as a vector named "vec"

```{r}

```

Test run this below (if TRUE -> continue with lab)

```{r}


```

## Read in data from a file

Usually, when we work with genomic data, the data comes in the form of a table. For example, we run a sequencing machine that outputs a fastq file with all the reads. After mapping/alignment to a genome, the sequencing reads can be "counted" (i.e. we have this many copies of gene A, and this many copies of gene B). This count data is in a matrix format, just like the data frame we tested above, with genes as rows and samples as columns. Lets say we made all the steps running up to us having the count file. Now, we want to read this into R so we can analyze it. 
First, download the count data from canvas via this [link](). Put the file in your current working directory and load it into R:

```{r}

counts <- read.table("tung/reads.txt", sep = "\t")

```

How many genes and samples are there in this data set? (HINT: e.g. use `dim()` or `nrows()`)

Which 5 genes are expressed the most across all samples?



## Plotting

a basic scatterplot is done via the `plot()` function.

```{r}

# Define the vector with 5 values
c <- c(1, 3, 6, 4, 9)

# Graph the  vector with all defaults
plot(c)

# Define 2 vectors
c1 <- c(1, 3, 6, 4, 9)
c2 <- c(2, 5, 4, 5, 12)

# Graph c1 using a y axis that ranges from 0 to 12
plot(c1, type="o", col="blue", ylim=c(0,12))

# Graph c2 with red dashed line and square points
lines(c2, type="o", pch=22, lty=2, col="red")

# Create a title with a red, bold/italic font
title(main="Two Lines", col.main="black")


#Histogram

hist(c1)

```

Lets plot our count data.

```{r}

#Distributions of gene counts
hist(log(rowSums(counts)))


```

What are we seeing here? Explain the plot!



## R/Bioconductor (Skip if already done)

Bioconductor is a repository of R-packages specifically for biological analyses. It has the strictest requirements for submission, including installation on every platform and full documentation with a tutorial (called a vignette) explaining how the package should be used. 

A "collection of bioinformatic tools" in summary. 

To install Bioconductor, run this in R:

```{r, eval=FALSE}

if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()

```

Then, there are **a lot** of "packages" within bioconductor that you can install and use. To install such a package you use type `BiocManager::install()` and within install you type the name of the package, e.g. `BiocManager::install("SummarizedExperiment")`

# Working with SummarizedExperiment

The SummarizedExperiment class is used to store rectangular matrices of experimental results, which are commonly produced by sequencing experiments.
Each object stores observations of one or more samples, along with additional meta-data describing both the observations (features) and samples (phenotypes).

A key aspect of the SummarizedExperiment class is the coordination of the meta-data and assays when sub-setting. For example, if you want to exclude a given sample you can do for both the meta-data and assay in one operation, which ensures the meta-data and observed data will remain in sync. Improperly accounting for meta and observational data has resulted in a number of incorrect results and retractions so this is a very desirable property.

Lets try it out!

We begin by loading the necessary packages, the `SummarizedExperiment` class is contained in the package with the name `SummarizedExperiment` (doh). Go ahead and load it

```{r, message=FALSE, warning=FALSE}

library(SummarizedExperiment)

```

There is an example data set included in the airway package, install/load this package first (if not installed/loaded):

```{r, eval=FALSE}


```

```{r}
library(airway)
data(airway)
se <- airway
se


```

How many genes do your SE have? How many samples?

Within the summarizedExperiment, you can store all kinds of data. type `colData(se)` to view all sample data included. Type `rowData(se)` to view all gene data.

What kind of sample data can you find?

To access a particular sample data, just type e.g. `se$cell` to find the cell data. 

You can also store meta data for the entire data set, type `metadata(se)` to view it. 



====================================================================================

# Appendix (If there is time and interest)

My preferred way of plotting is to use the very flexible "ggplot" package. Getting good at plotting will allow you to make some quite cool plots as well as animations (check: https://www.tylermw.com/3d-ggplots-with-rayshader/) and interactive plots. We will use plotting for plotting in the later labs, but to get familiar with the concept, run a bit in this tutorial and play around with it (try e.g. to use it on the airway data, or our own count data):

https://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html


END

